gensim/models/base_any2vec.py

```diff
@@ -0,0 +1,781 @@
+            trained_word_count += trained_word_count_epoch
+            raw_word_count += raw_word_count_epoch
+            job_tally += job_tally_epoch
+
+            for callback in self.callbacks:
+                callback.on_epoch_end(self)
+
+        # Log overall time
+        total_elapsed = default_timer() - start
+        self._log_train_end(raw_word_count, trained_word_count, total_elapsed, job_tally)
+
+        self.train_count += 1  # number of times train() has been called
+        self._clear_post_train()
+
+        for callback in self.callbacks:
+            callback.on_train_end(self)
+        return trained_word_count, raw_word_count
+
+    @classmethod
+    def load(cls, fname_or_handle, **kwargs):
+        return super(BaseAny2VecModel, cls).load(fname_or_handle, **kwargs)
+
+    def save(self, fname_or_handle, **kwargs):
+        super(BaseAny2VecModel, self).save(fname_or_handle, **kwargs)
+
+
+class BaseVocabBuilder(utils.SaveLoad):
+    """Class for managing vocabulary of a model. Takes care of building, pruning and updating vocabulary."""
+    def __init__(self):
+        self.vocab = {}
+        self.index2word = []
+
+    def scan_vocab(self, data_iterable, progress_per=10000, **kwargs):
+        """Do an initial scan of all words appearing in data_iterable.
+        Sets num_examples(total examples in data_iterable) and
+        raw_vocab(collections.defaultdict(int) mapping str vocab elements to their counts for all vocab words)"""
+        raise NotImplementedError
+
+    def prepare_vocab(self, update=False, **kwargs):
+        raise NotImplementedError
+
+
+class BaseModelTrainables(utils.SaveLoad):
+    """Class for storing and initializing/updating the trainable weights of a model. Also includes
+    tables required for training weights. """
+    def __init__(self, vector_size=None, seed=1):
+        self.vectors = []
+        self.vector_size = int(vector_size)
+        self.seed = seed
+
+    def prepare_weights(self, update=False, vocabulary=None):
+        raise NotImplementedError
+
+    def reset_weights(self, vocab):
+        """Reset all trainable weights to an initial (untrained) state, but keep the existing vocabulary."""
+        raise NotImplementedError
+
+    def update_weights(self, vocab):
+        """
+        Copy all the existing weights, and reset the weights for the newly
+        added vocabulary.
+        """
+        raise NotImplementedError
+
+    def seeded_vector(self, seed_string):
+        """Create one 'random' vector (but deterministic by seed_string)"""
+        raise NotImplementedError
+
+
+class BaseWordEmbedddingsModel(BaseAny2VecModel):
+    def __init__(self, sentences=None, workers=3, vector_size=100, epochs=5, callbacks=(), batch_words=10000,
+                 trim_rule=None, sg=0, alpha=0.025, window=5, seed=1, hs=0, negative=5, cbow_mean=1,
+                 min_alpha=0.0001, compute_loss=False, **kwargs):
+        self.sg = int(sg)
```

Comment:
It would be nice to be consistent about the type conversion. If we convert `sg` to `int`, we should probably convert `hs` and `negative` to `int`, too, and convert `compute_loss` to `bool`. Ditto elsewhere.